s3auth-relay/src/main/java/com/s3auth/relay/AbstractFacade.java

```diff
@@ -0,0 +1,292 @@
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.TimeUnit;
+import lombok.EqualsAndHashCode;
+import lombok.ToString;
+
+/**
+ * Abstract facade (port listener).
+ *
+ * @author Yegor Bugayenko (yegor@tpc2.com)
+ * @author Simon Njenga (simtuje@gmail.com)
+ * @version $Id$
+ * @since 0.0.1
+ */
+@ToString
+@EqualsAndHashCode(of = { "sockets", "server", "secured" })
+@SuppressWarnings({ "PMD.DoNotUseThreads", "PMD.TooManyMethods" })
+@Loggable(Loggable.DEBUG)
+abstract class AbstractFacade implements Closeable {
+
+    /**
+     * How many threads to use.
+     */
+    protected static final int THREADS = Tv.HUNDRED;
+
+    /**
+     * Executor service, with socket openers.
+     */
+    private transient ScheduledExecutorService frontend;
+
+    /**
+     * Executor service, with consuming threads.
+     */
+    private transient ScheduledExecutorService backend;
+
+    /**
+     * Blocking queue of ready-to-be-processed sockets.
+     */
+    private final transient BlockingQueue<Socket> sockets =
+        new SynchronousQueue<Socket>();
+
+    /**
+     * Server socket.
+     */
+    private transient ServerSocket server;
+
+    /**
+     * Secured Server socket.
+     */
+    private transient ServerSocket secured;
+
+    @Override
+    public void close() throws IOException {
+        try {
+            this.shutdown(this.getFrontend());
+            this.shutdown(this.getBackend());
+        } catch (final InterruptedException ex) {
+            Thread.currentThread().interrupt();
+            throw new IOException(ex);
+        }
+        this.getServer().close();
+        this.getSecured().close();
+    }
+
+    /**
+     * Get frontend.
+     * @return The frontend
+     */
+    protected final ScheduledExecutorService getFrontend() {
+        return this.frontend;
+    }
+
+    /**
+     * Set frontend.
+     * @param frntend The frontend to set
+     */
+    protected final void setFrontend(final ScheduledExecutorService frntend) {
+        this.frontend = frntend;
+    }
+
+    /**
+     * Get backend.
+     * @return The backend
+     */
+    protected final ScheduledExecutorService getBackend() {
+        return this.backend;
+    }
+
+    /**
+     * Set backend.
+     * @param bckend The backend to set
+     */
+    protected final void setBackend(final ScheduledExecutorService bckend) {
+        this.backend = bckend;
+    }
+
+    /**
+     * Get server socket.
+     * @return The server
+     */
+    protected final ServerSocket getServer() {
+        return this.server;
+    }
+
+    /**
+     * Set server socket.
+     * @param svr The server socket to set
+     */
+    protected final void setServer(final ServerSocket svr) {
+        this.server = svr;
+    }
+
+    /**
+     * Get secured server socket.
+     * @return The secured server
+     */
+    protected final ServerSocket getSecured() {
```

Comment:
@simonjenga no getters/setters. use a constructor and set the members as final to keep it immutable.